### PR TITLE
Lcov testing

### PR DESCRIPTION
**Pull request by bradb423**
Hi @nedbat. As a follow up to our messages last week, we have produced an LCOV report generator, taking inspiration from @chaychoong's [cli converter](https://github.com/chaychoong/coveragepy-lcov).
The LCOV generator is run similarly to the HTML or JSON generators using the command `coverage lcov`, and produces an LCOV file (defaulting to `coverage.lcov`).

As of right now, the LCOV reporter does not support branch coverage, however this feature has the potential to be added in the future.

The generator also provides checksums for each line in the `coverage.lcov` file, as outlined [here](http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php)

### CODE REVIEW

File: tests/test_lcov.py
Comment (TLATER) [inline tests/test_lcov.py]: Hmm, I thought the issue was that you didn't think this was possible with coveragepy because none of the other examples tried it?

It's worth noting that lcov [has two different kinds of branch coverage reporting](https://github.com/linux-test-project/lcov/blob/master/man/geninfo.1#L552):

1. Per-line "is this branch covered"
2. Summary of branch coverage

Having had a quick glance, the latter shouldn't be all that hard to do, see https://github.com/nedbat/coveragepy/blob/1c29ef3e4b871508bb2defd6b47b9a745547b626/coverage/results.py#L140 for getting the data.

I think you'd simply need to add something like:

```python
brf, brh = analysis.branch_stats()
outfile.write(f"BRF:{brf}\n")
outfile.write(f"BRH:{brh}\n")
```

The former can be done with [`arcs`](https://github.com/nedbat/coveragepy/blob/1c29ef3e4b871508bb2defd6b47b9a745547b626/coverage/results.py#L79), I believe, but I don't know enough about what those represent to actually recommend anything here without playing a bit with it myself.

I'm guessing arcs don't line up perfectly with lines (e.g. `something if cond else something_else`), and that this therefore requires a bit of thought, but this shouldn't be a complete show-stopper, especially since C also has the option of using blocks in lines around branches?

File: coverage/lcovreport.py
Comment (TLATER) [inline coverage/lcovreport.py]: Besides branch coverage, [function name and execution info](https://github.com/linux-test-project/lcov/blob/master/man/geninfo.1#L538) is also missing. This is totally valid lcov, but it's probably worth pointing out in a comment here at least.

I don't think there's a good way of implementing that with coveragepy currently?

**Comment by bradb423** [comment]
I have added the branch coverage support for the LCOV generator. Both the summary, and the line by line branch coverage were possible in the end, and appropriate tests for branch coverage have now been added. 

There is still the problem of pypy3 failing currently, this seems to be due to the ordering of the lines in the LCOV file, a further patch will be added to rectify this.

**Comment by bradb423** [comment]
I have now fixed the pypy3 issue from earlier.

**Comment by nedbat** [comment]
I really appreciate the work you are putting into this.  I'm holding off on specific review while I continue to think through where this code should live (inside coverage.py or outside, roughly).

**Comment by bradb423** [comment]
Thank you for your time, @nedbat. I'd be more than happy to answer any questions that you may have, along with any review comments that you may have in the future.

From what I can tell, there seems to be a bug in the test for pypy3 on MacOS. It looks like the pypy3 cannot be found for the run, as you can see [here](https://github.com/nedbat/coveragepy/runs/4519927975?check_suite_focus=true), do you know what might be the cause of this?

**Comment by nedbat** [comment]
@bradb423 I fixed the pypy3 thing on master. If you rebase, you'll get the fix.

**Comment by bradb423** [comment]
@nedbat thank you!

**Comment by bradb423** [comment]
Hi @nedbat , thank you for your patience and for your help over the last couple of weeks. I encountered a bug with the code this morning (causing the block numbers to provide the wrong values) which has now been fixed.

I believe that as of right now, this PR is in a state at which it does not require any more commits. Please let me know your thoughts, and your questions if you have any.

File: coverage/cmdline.py
Comment (nedbat) [inline coverage/cmdline.py]: This should be `-o --output` I think.  There's an earlier PR that used `-l`, but that was hacking it into an existing command.

File: coverage/cmdline.py
Comment (nedbat) [inline coverage/cmdline.py]: I've tried to keep the options alphabetized (roughly) in the lists.

File: coverage/cmdline.py
Comment (nedbat) [inline coverage/cmdline.py]: I thought this pre-assignment was odd, but I see now that the other reporting clauses do it like that.  I'm changing them :)

File: coverage/config.py
Comment (nedbat) [inline coverage/config.py]: This comment is a little misleading, since there is no parsing of an `[lcov]` section in the config file.  Do you want to support setting the lcov output file in the config file?

File: coverage/control.py
Comment (nedbat) [inline coverage/control.py:1062]: This docstring should have:
```
.. versionadded:: 6.3
```

File: coverage/lcovreport.py
Comment (nedbat) [inline coverage/lcovreport.py]: This is a super-nit, but I prefer that docstrings and comment use sentence punctuation, with a final period.

File: coverage/lcovreport.py
Comment (nedbat) [inline coverage/lcovreport.py]: No need for `list()` here.

File: coverage/lcovreport.py
Comment (nedbat) [inline coverage/lcovreport.py]: According to the commit comment, this except is only for the case of an empty file? If so, it would be better to use an explicit `if sourceLines:` test.

File: coverage/lcovreport.py
Comment (nedbat) [inline coverage/lcovreport.py]: `sourceLines` is an unusual variable name.  Can we use `source_lines` instead?

File: coverage/lcovreport.py
Comment (nedbat) [inline coverage/lcovreport.py:61]: I guess tastes vary, but I like `.strip("=")` here more than `[:-2]` since it's more explicit what it's doing.
And "ascii" as encoding also seems more appropriate.

Comment (nedbat) [reply coverage/lcovreport.py:61]: But this is where we need to decide what to do about md5: it's not available on FIPS-enabled systems.  Here are some choices:
1. Don't use md5 at all, and leave the hash out of the report.
2. Use md5, but catch the ValueError it will raise on FIPS systems (https://github.com/nedbat/coveragepy/issues/1189)
3. Make lcov unavailable on FIPS systems
4. Let lcov just fail if used on FIPS systems

I don't know what the effect would be of omitting the hash, so I don't know which to pick.

**Review by nedbat — CHANGES_REQUESTED** [review]
Thanks for pushing forward on this feature. I haven't review the whole pull request, but wanted to get you some feedback. More tomorrow.

File: tests/test_lcov.py
Comment (nedbat) [inline tests/test_lcov.py]: `LcovTest` :)

File: tests/test_lcov.py
Comment (nedbat) [inline tests/test_lcov.py]: This function doesn't make an lcov report as it claims to.

File: tests/test_lcov.py
Comment (nedbat) [inline tests/test_lcov.py]: This function doesn't have a module parameter.

File: tests/test_lcov.py
Comment (nedbat) [inline tests/test_lcov.py]: I'm not a fan of returning a list of lines here.  I think the tests will read better if this function returns the text as one string, and the literals are multi-line strings instead of lists of lines.

File: tests/test_lcov.py
Comment (nedbat) [inline tests/test_lcov.py]: Shouldn't the main_file be run by the test_file, instead of directly?

**Review by nedbat — CHANGES_REQUESTED** [review]
OK, I've read the whole thing.  We'll also need the docs updated, if you want to do that.

File: coverage/lcovreport.py
Comment (TLATER) [reply coverage/lcovreport.py:61]: The hash is mainly used to ensure the correct lines are marked as covered, to my knowledge. Upstream's `genhtml`, for example, which uses an lcov file and the original source tree to produce an html page with annotated source files, will throw an error if it finds a hash that mismatches: https://github.com/linux-test-project/lcov/blob/master/bin/genhtml#L5308.

Genhtml is of course an unlikely usecase with pycoverage, but it's the reference consumer.

It is up to the implementation of the consumer to define what the md5 is used for, but it is marked as optional, so any tools that break on a missing md5 are inherently broken. There shouldn't be any issue with simply omitting the hash when it can't be set, and I can't think of many other uses than "has our user accidentally changed their source". I think that was probably a more common use case around the time lcov added that feature :)

It *would* be sad if pycoverage produced different outputs depending on the machine it is run on. Maybe hence it's best to remove this functionality in general.

I think a fifth option is to lobby upstream for supporting a more modern hash, as much as that seems unnecessary. It's honestly a bit silly that python mandates thinking about FIPS compliance with no opt-out if you use its md5 hash, and even more silly that some default python distributions seem to enable FIPS compliance (this seems to be a MacOS thing?).

It'll be hard to come up with a backwards-compatible design for that, so I suspect the solution would be to deprecate having hashes at all, since those don't really seem that useful anymore in 2022 when checking for accidental source changes is anyway so easy and often unnecessary.

Comment (nedbat) [reply coverage/lcovreport.py:61]: Thanks for the backstory on the hash values.

> It's honestly a bit silly that python mandates thinking about FIPS compliance with no opt-out if you use its md5 hash

Is this Python's fault? FIPS forbids md5 as an algorithm.

> some default python distributions seem to enable FIPS compliance (this seems to be a MacOS thing?)

I haven't encountered this.  Can you provide more information?

Comment (TLATER) [reply coverage/lcovreport.py:61]: > Is this Python's fault? FIPS forbids md5 as an algorithm.

No, not at all. I just think the situation is a bit silly for backwards-compatibility with older formats that use md5 for legitimate reasons (or even just for migration purposes). It's besides the point, sorry for venting here.

> I haven't encountered this. Can you provide more information?

[This support page](https://support.apple.com/en-us/HT209638) seems to suggest a FIPS-compliance mode is enabled by default. Reading further, it may come down to processor support for the hash, rather than something explicitly disabled on the software side, but this is all ultimately besides the point.

I think I'll spend a bit more time reading up on where exactly the limitation comes from, and then raise an issue on the lcov upstream to see if they're interested enough in allowing lcov to be run in FIPS-compliant environments. No need for coveragepy to be involved in that :)

**Comment by nedbat** [comment]
btw, for extra credit, it would be good to have tests in test_cmdline.py also.

File: tests/test_lcov.py
Comment (bradb423) [reply tests/test_lcov.py]: Hi there,

Is there a way to do the above? I am just about finished with making the changes, but this change in the tests is stumping me at the moment.

Cheers,

Brad

Comment (nedbat) [reply tests/test_lcov.py]: Sorry, I'm not sure what you are asking. My point was simply that you don't need to call `start_import_stop` on both the main_file and the test_file.  test_file.py imports main_file, so just calling it on test_file will be enough, and is how coverage would normally be run.

Comment (bradb423) [reply tests/test_lcov.py]: Thank you, it seems I misunderstood the review comment. I have amended the tests now, thank you!

File: coverage/lcovreport.py
Comment (bradb423) [reply coverage/lcovreport.py:61]: @TLATER is right here on the use of the hash within LCOV. I am perfectly happy to remove the md5 hash if you would prefer @nedbat.

**Comment by nedbat** [comment]
@bradb423 Good to see progress! Let me know when you are ready.

**Comment by bradb423** [comment]
Hi @nedbat Thanks for your patience. I'd say that this is ready for another look. Please let me know how you feel about the current state, along with the md5 hash :)

**Comment by nedbat** [comment]
Thanks, this looks good to merge.  We'll keep the md5 as it is, and see if we get any concerns from people.  I'll do some styling and clean-up after the merge.

Thanks for doing this! <3

**Review by nedbat — APPROVED** [review]

**Comment by nedbat** [comment]
This is now released as part of [coverage 6.3](https://pypi.org/project/coverage/6.3).  Thanks again! <3

**Pull request merged**
